civismlext/preprocessing.py

```diff
@@ -141,12 +146,23 @@ def __init__(self,
                  cols_to_expand='auto',
                  dummy_na=True,
                  fill_value=0.0,
-                 dataframe_output=False):
+                 dataframe_output=False,
+                 drop_null_cols=False):
         self.cols_to_drop = cols_to_drop
         self.cols_to_expand = cols_to_expand
         self.dummy_na = dummy_na
         self.fill_value = fill_value
         self.dataframe_output = dataframe_output
+        self.drop_null_cols = drop_null_cols
+
+    def _flag_nulls(self, X, cols_to_drop):
+        null_cols = [col for col in X if
+                     col not in cols_to_drop and X[col].isnull().all()]
```

Comment:
Looking at this line -- you don't really care about the null status of every entry in the DataFrame. All you want to know is whether or not there's at least one non-null value. Therefore, I have two suggestions for performance improvements.

First, I've found a significant improvement in run time by replacing `X[col].isnull().all()` with `X[col].first_valid_index() is None`. It's 4x faster for a tiny DataFrame, and 10x faster for a 1M row DataFrame.

Second, if you think the normal use case is for most columns to be mostly non-null (I think that's the case), you can add a further condition to the list comprehension:
```python
null_cols = [col for col in X if
                    col not in cols_to_drop and 
                    pd.isnull(X[col].values[0]) and
                    X[col].first_valid_index() is None]
```
Checking if the first element of the series is null takes about 4 us when I test it locally. That's significantly faster than either the `isnull` check or the `first_valid_index` check.

Reply:
You'll need to special-case empty `DataFrame`s when using these checks. You should skip the null check if the `DataFrame` length is zero.

Reply:
I like this idea a lot-- really clever idea, to check the first value to save extra time!